src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -0,0 +1,181 @@
+
+
+class EnvironmentMetricsService(protocol.ServerSlice):
+    """Slice for the management of metrics"""
+
+    def __init__(self) -> None:
+        super(EnvironmentMetricsService, self).__init__(SLICE_ENVIRONMENT_METRICS)
+        self.metrics_collectors: Dict[str, MetricsCollector] = {}
+        self.previous_timestamp = datetime.now()
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> List[str]:
+        return [SLICE_TRANSPORT]
+
+    async def start(self) -> None:
+        await super().start()
+        self.schedule(self.flush_metrics, COLLECTION_INTERVAL_IN_SEC, initial_delay=0, cancel_on_stop=True)
+
+    def register_metric_collector(self, metrics_collector: MetricsCollector) -> None:
+        """
+        Register the given metrics_collector.
+        """
+        if metrics_collector.get_metric_name() not in self.metrics_collectors:
+            self.metrics_collectors[metrics_collector.get_metric_name()] = metrics_collector
+        else:
+            raise Exception(f"There already is a metric collector with the name {metrics_collector.get_metric_name()}")
+
+    async def flush_metrics(self) -> None:
+        """
+        Invoked at the end of the metrics collection interval. Writes the metrics
+        collected by the MetricsCollectors in the past metrics collection interval
+        to the database.
+        """
+        now: datetime = datetime.now()
+        old_previous_timestamp = self.previous_timestamp
+        self.previous_timestamp = now
+        metric_gauge: Sequence[EnvironmentMetricsGauge] = []
+        metric_timer: Sequence[EnvironmentMetricsTimer] = []
+
+        def create_metrics_gauge(metric_values_gauge: Sequence[MetricValue], timestamp: datetime):
+            for mv in metric_values_gauge:
+                metric_gauge.append(EnvironmentMetricsGauge(metric_name=mv.metric_name, timestamp=timestamp, count=mv.count))
+
+        def create_metrics_timer(metric_values_timer: Sequence[MetricValueTimer], timestamp: datetime):
+            for mv in metric_values_timer:
+                metric_timer.append(
+                    EnvironmentMetricsTimer(metric_name=mv.metric_name, timestamp=timestamp, count=mv.count, value=mv.value)
+                )
+
+        async with EnvironmentMetricsGauge.get_connection() as con:
+            for mc in self.metrics_collectors:
+                metrics_collector: MetricsCollector = self.metrics_collectors[mc]
+                metric_type: str = metrics_collector.get_metric_type()
+                metric_values: Sequence[MetricValue] = await metrics_collector.get_metric_value(
+                    old_previous_timestamp, now, connection=con
+                )
+                if metric_type == MetricType.GAUGE:
+                    create_metrics_gauge(metric_values, now)
+                elif metric_type == MetricType.TIMER:
+                    create_metrics_timer(metric_values, now)
```

Comment:
@sanderr it is here that I'm having trouble with the typing: as reminder `metric_values` is of type `Sequence[MetricValue]`. But once we get here it should always be of type `Sequence[MetricValueTimer]` and I'd like to enforce that.